File: libraries/chr/chr_putc.f
      SUBROUTINE CHR_PUTC( STR1, STR2, IPOSN )
*+
*  Name:
*     CHR_PUTC

*  Purpose:
*     Put a CHARACTER string into another at a given position.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL CHR_PUTC( STR1, STR2, IPOSN )

*  Description:
*     The string STR1 (or as much of it as there is room for) is
*     copied into the part of STR2 beginning at position IPOSN+1.
*     IPOSN is updated to indicate the end position of the copy of
*     STR1 within STR2 after this operation. If no copying is
*     done, IPOSN is returned unchanged. The sizes of STR1 and
*     STR2 are based on the declared Fortran 77 size given by the
*     intrinsic function LEN.

*  Arguments:
*     STR1 = CHARACTER * ( * ) (Given)
*        The string to be copied.
*     STR2 = CHARACTER * ( * ) (Given and Returned)
*        The string into which STR1 is to be copied.
*     IPOSN = INTEGER (Given and Returned)
*        The position pointer within STR2.

*  Copyright:
*     Copyright (C) 1983, 1984, 1988, 1989 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     JRG: Jack Giddings (UCL)
*     ACD: A.C. Davenhall (ROE)
*     AJC: A.J. Chipperfield (STARLINK)
*     {enter_new_authors_here}

*  History:
*     3-JAN-1983 (JRG):
*        Original version.
*     2-OCT-1984 (ACD):
*        Documenation improved.
*     13-SEP-1988 (AJC):
*        Documentation improved.
*     24-FEB-1989 (AJC):
*        Check on string sizes.
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Arguments Given:
      CHARACTER STR1 * ( * )

*  Arguments Given and Returned:
      CHARACTER STR2 * ( * )

      INTEGER IPOSN

*  Local Variables:
      INTEGER SIZE1              ! Size of STR1
      INTEGER SIZE2              ! Size of STR2

*.

*  Get the size of target string.
      SIZE2 = LEN( STR2 )

*  Check that the pointer is within string.
      IF ( IPOSN .LT. SIZE2 ) THEN

*     Get the length that can be copied.
         SIZE1 = MIN( LEN( STR1 ), SIZE2-IPOSN )

         IF ( SIZE1 .GT. 0 ) THEN

*        Copy the string.
            STR2( IPOSN+1 : IPOSN+SIZE1 ) = STR1( 1 : SIZE1 )

*        Update the pointer value.
            IPOSN = IPOSN + SIZE1
         END IF
      END IF

      END

File: libraries/chr/chrPutc.c
#include "chr.h"
#include "star/util.h"
#include <string.h>

void chrPutc( const char *str1, char *str2, size_t str2_length, size_t *iposn ){
/*
*+
*  Name:
*     chrPutc

*  Purpose:
*     Put a character string into another at a given position.

*  Synopsis:
*     void chrPutc( const char *str1, char *str2, size_t str2_length,
*                   size_t *iposn )

*  Description:
*     The string "str1" (or as much of it as there is room for) is copied
*     into the part of "str2" beginning at position "iposn+1". "iposn" is
*     updated to indicate the end position of the copy of "str1" within
*     "str2" after this operation. If no copying is done, "iposn" is
*     returned unchanged.

*  Parameters:
*     str1
*        Pointer to a null terminated string holding the string to be
*        copied.
*     str2
*        Pointer to a null terminated string holding the string into which
*        "str1" is to be copied.
*     str2_length
*        The declared length of the supplied 'str2' array. This should include
*        room for the terminating null.
*     *iposn
*        The zero-based position pointer within "str2".

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     DSB: David S. Berry (EAO)

*  History:
*     15-MAY-2018 (DSB):
*        Original version, based on equivalent Fortran function.

*-
*/

/* Local Variables: */
   size_t size1;         /* Size of STR1 */
   size_t size2;         /* Size of STR2 */

/* Check the supplied string length is OK. */
   if( str2_length > 0 ) {

/* Get the max size of target string, excluding the trailing null. */
      size2 = str2_length - 1;

/* Check that the pointer is within string. */
      if( *iposn < size2 ) {

/* Get the length that can be copied. */
         size1 = strlen( str1 );
         size2 -= *iposn;
         if( size2 < size1 ) size1 = size2;

/* Copy the string. */
         if( size1 > 0 ) {
            star_strlcpy( str2 + *iposn, str1, size1 + 1 );

/* Update the pointer value. */
            *iposn += size1;
         }
      }
   }
}

